Implement minimal TCP module ("net")

matiasinsaurralde opened the pull request:
Similar to #209 

I've added tests and also a sample echo server in `samples/echoserver.ts`.

jedahan commented on msg.proto:
```diff
@@ -99,4 +108,11 @@ message Msg {
   uint32 write_file_sync_perm = 132;
   // write_file_sync_perm specified by https://godoc.org/os#FileMode
+
+  int32 net_port = 134;
```
What is the reason this is 134 instead of 133?

matiasinsaurralde replied:
Fixed

matiasinsaurralde commented on net.ts:
```diff
@@ -0,0 +1,149 @@
+export class NetServerConn {
+    private readonly id: number;
+    private onDataCb: (data: Uint8Array) => void;
+    constructor(id: number) {
```
👍

yordis commented on deno.d.ts:
```diff
@@ -11,4 +12,38 @@ declare module "deno" {
-}
+
+  export class NetSocket {
+    private connectCb: () => void;
```
`onConnect` maybe?

yordis commented on deno.d.ts:
```diff
@@ -11,4 +12,38 @@ declare module "deno" {
+
+  export class NetSocket {
+    private connectCb: () => void;
+    private onDataCb: (data: Uint8Array) => void;
```
`onData` without the `Cb`. It reads the same.

gillchristian commented on net.go:
```diff
@@ -0,0 +1,167 @@
+func netClientClose(msg *Msg) {
+	client, ok := clients[msg.NetClientId]
+	if !ok {
+	}
```
Why is this block empty? Is it just WIP? (same for `netClientWrite` above)

gillchristian commented on net.go:
```diff
@@ -0,0 +1,167 @@
+	for {
+		data, _, err := reader.ReadLine()
+		if err != nil {
+			return
```
On [`netSocketRead`](https://github.com/matiasinsaurralde/deno/blob/2b59bf260313628eb3b0b39ebd4fd94b960e71ea/net.go#L85-L99) you `break` on error when reading line. Any reason for a `return` here, shouldn't it also be `break` here?

gillchristian commented on tests.ts:
```diff
@@ -55,3 +57,30 @@ test(async function tests_writeFileSync() {
+  const decoder = new TextDecoder("utf-8");
+  const server = createServer((conn: NetServerConn) => {
+    // console.log("conn writes");
+    // conn.write("welcome\n");
```
Maybe remove this commented code?

matiasinsaurralde commented:
@gillchristian Thanks for the comments, just updated the code

ry commented:
@matiasinsaurralde Awesome!! Looks quite good. I will review in more detail soon. (Very sorry for the delays.)

ry commented:
@matiasinsaurralde Let's land your http patch (#209 ) first. Many of my comments in that PR will apply here.

matiasinsaurralde commented:
Just FYI: have refactored the messages a bit, might still need tweaks.
Also added the `wg` calls for proper blocking when the server runs.
Added the `close` method in both TS and Go but the implementation is missing, I've added a `TODO` comment, could be for another PR.

ry commented:
@matiasinsaurralde We will come back to this soon but I'm closing this PR for now.
